// bootstrap-vue-next/dist/src/directives/BTooltip/index.mjs
import "vue";
import { d as unbind, e as resolveActiveStatus, g as resolveContent, h as bind, j as resolveDirectiveProps } from "../../../floatingUi-CE0Z1avc.mjs";
import { f as defaultsKey } from "../../../keys-BLeKMItg.mjs";
import { f as findProvides } from "../../../utils-DAHbMzJg.mjs";
const vBTooltip = {
  mounted(el, binding, vnode) {
    const defaults = findProvides(binding, vnode)[defaultsKey]?.value;
    const isActive = resolveActiveStatus(binding.value);
    if (!isActive) return;
    const text = resolveContent(binding.value, el);
    if (!text.body && !text.title) return;
    el.$__binding = JSON.stringify([binding.modifiers, binding.value]);
    bind(el, binding, {
      noninteractive: true,
      ...defaults["BTooltip"] || void 0,
      ...resolveDirectiveProps(binding, el),
      title: text.title ?? text.body ?? "",
      tooltip: isActive
    });
  },
  updated(el, binding, vnode) {
    const defaults = findProvides(binding, vnode)[defaultsKey]?.value;
    const isActive = resolveActiveStatus(binding.value);
    if (!isActive) return;
    const text = resolveContent(binding.value, el);
    if (!text.body && !text.title) return;
    delete binding.oldValue;
    if (el.$__binding === JSON.stringify([binding.modifiers, binding.value])) return;
    unbind(el);
    bind(el, binding, {
      noninteractive: true,
      ...defaults["BTooltip"] || void 0,
      ...resolveDirectiveProps(binding, el),
      title: text.title ?? text.body ?? "",
      tooltip: isActive
    });
    el.$__binding = JSON.stringify([binding.modifiers, binding.value]);
  },
  beforeUnmount(el) {
    unbind(el);
  }
};
export {
  vBTooltip
};
